### PR TITLE
Explain exception logging, monitoring and alerts

### PR DESCRIPTION
**Pull request by bittner**
We get quite a lot of support requests asking for guidance on using application logging.

* Explains application logging and options for exception logging, monitoring and alerts
* Reformats troubleshooting chapter to match other documents, cross-link with new explanations

**Internal reference:** VINT-1906, APPU-2544, various customer tickets

### CODE REVIEW

File: app/customapplications.rst [hunk 1]
```diff
@@ -8,14 +10,61 @@ from source code.
...
+
+.. _Sentry: https://sentry.io/_/resources/customer-success/alert-rules/
+.. _New Relic: https://docs.newrelic.com/docs/alerts/new-relic-alerts/getting-started/introduction-new-relic-alerts
+.. _Datadog: https://docs.datadoghq.com/monitors/notifications/?tab=is_alert
+
+Monitoring & Alerting
+---------------------
+
+From a concept point of view, cloud applications should be *self-healing* and
+hence do not strictly need to be monitored. You should specify `resource
```
Comment (tobru) [inline app/customapplications.rst]: Not only requests and limits, but the also the probes should be mentioned, they're much more important.

File: troubleshooting.rst [hunk 2]
```diff
@@ -1,39 +1,54 @@
...
-If your app does not deploy or run have a look at some tips here. If you don't find your error or do find an error message you don't understand come to the community chat at https://community.appuio.ch and ask.
-
-The tips are structured in three categories: how to inspect the build process, how to inspect the deploy process, how to inspect your running application and some specific tipps and error messages at the end of each part.
+If your app does not deploy or run have a look at some tips here. If you don't
+find your error or do find an error message you don't understand come to the
+community chat at https://community.appuio.ch and ask.
 
+The tips are structured in three categories: how to inspect the build process,
+how to inspect the deploy process, how to inspect your running application and
+some specific tipps and error messages at the end of each part.
```
Comment (chamer81) [inline troubleshooting.rst]:
```suggestion
some specific tips and error messages at the end of each part.
```

File: troubleshooting.rst [hunk 3]
```diff
@@ -1,39 +1,54 @@
...
 
 Build
 -----
 
-Note that the build process of your application is just another pod that you can look at in the Web-GUI or CLI. The build process will use computing resources in your quota just like your application does, if there are no more available resources the build won't run.
-
+Note that the build process of your application is just another pod that you
+can look at in the Web GUI or CLI. The build process will use computing
+resources in your quota just like your application does, if there are no more
```
Comment (chamer81) [inline troubleshooting.rst]:
```suggestion
resources in your quota just like your application does, so if there are no more
```

**Comment by bittner** [comment]
Good feedback, thanks! – I incorporated the changes.

_That's one of those moments when I realize how valuable code review is!_

**Comment by tobru** [comment]
While I like the latest pushed version better than the version before, I have a bit of an issue with the statement

> and hence do not strictly need to be monitored

IMHO the suggested measurements do not help when f.e. when a Pod is in `CrashLoopBack`, you would still need monitoring to get notified about that. I'm not sure if we should really state that this way. Isn't there another way to inform users about the necessity of these measurements, leaving the "do not monitor" statement out?

**Comment by bittner** [comment]
> IMHO the suggested measurements do not help when f.e. when a Pod is in `CrashLoopBack`, you would still need monitoring to get notified about that.

I understand your concern. The statement is merely a bold statement to set "modern approaches" in contrast with the traditional manual releasing and application babysitting practices of traditional operations. We agreed on that to be valuable to mention with @ccremer.

I'll try to come up with a less provocative statement that is closer to reality. True. :+1:

**Comment by ccremer** [comment]
@bittner That is not what we discussed.

Applications still need active monitoring IMO, it's just that certain monitoring checks do not make sense **for VSHN**. If there are exceptions happening in the application, we as VSHN can not do more than "Dear customer, your app has errors", as we cannot fix them simply because we don't know the application.

So, CrashLoopBackoffs are still ok to be monitored, as that is something we can sometimes fix. It can be that liveness checks are configured wrong, resource limits too restricted, etc.

The ideal case IMO is when the customer has setup a prometheus with custom alerts, where the alerts are sent to people who can do something about it.

**Review by chamer81 — APPROVED** [review]
I agree with the remarks by colleagues about monitoring alerts. Aside from that, it looks good.

**Comment by bittner** [comment]
> If there are exceptions happening in the application, we as VSHN can not do more than

That is what application monitoring, aka "Exception Tracking", is about. That is covered by the change above.

I have made the paragraph on "Monitoring & Alerting" turned into a constructive suggestion on what action to take. I think that should be appealing to everyone now. Please, check again!

File: app/customapplications.rst [hunk 4]
```diff
@@ -8,14 +10,66 @@ from source code.
...
+and hence have less need to be monitored for everything. You should specify
+`resource requests and limits`_ for your Pods as well as `liveness and
+readiness probes`_, then APPUiO can automatically kill and restart your
+applications when they start using too much CPU or memory, or are stuck in
+a broken state. In addition, we suggest the following options to actively
+monitor your application and receive alerts:
+
+- **Prometheus (with AlertManager)**, self-hosted. It is best installed using
+  a `Helm chart`_. See the Prometheus docs on how to configure
+  `alert backends`_ and `alerting rules`_.
```
Comment (ccremer) [inline app/customapplications.rst]: The upstream Prometheus Helm Chart is not Openshift compatible, it doesn't make sense to link it.

Comment (bittner) [reply app/customapplications.rst]: Interesting. That was exactly your proposal, though. So, which one should I link to instead?

Comment (ccremer) [reply app/customapplications.rst]: Yes, until @corvus-ch wrote in Chat that it doesn't work on Openshift.
I do not know. The current workaround seems to be to create a new repository out from a `helm template`, fix the incompatible stuff and add more RBAC.

Comment (srueg) [reply app/customapplications.rst]: Maybe link the Prometheus operator?
It's not yet available on APPUiO for all users, but a customer could deploy his own operator for his own namespace...

Comment (bittner) [reply app/customapplications.rst]: Sounds like, good that we talk about a problem that needs to be solved for our customers! :smirk:

Comment (bittner) [reply app/customapplications.rst]: > Yes, until @corvus-ch wrote in Chat that it doesn't work on Openshift.

Seems like a problem that only affects parts of the helm chart, not the chart in general. https://github.com/helm/charts/issues/20797 (closed only due to inactivity) Or maybe not. https://github.com/helm/charts/issues/16319#issuecomment-521630629  :thinking:

Comment (bittner) [reply app/customapplications.rst]: Discussed this internally with @corvus-ch and adapted the text accordingly. Until we can offer our customers a self-service solution we'll simply let them ask us for help.

This PR should now be ready for merging.

**Comment by tobru** [comment]
Thanks, I've merged it now :+1:

**Pull request merged**